test/test_cmd_vel_mux.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+# All rights reserved.
+#
+# Software License Agreement (BSD License 2.0)
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+# * Redistributions of source code must retain the above copyright
+#   notice, this list of conditions and the following disclaimer.
+# * Redistributions in binary form must reproduce the above
+#   copyright notice, this list of conditions and the following
+#   disclaimer in the documentation and/or other materials provided
+#   with the distribution.
+# * Neither the name of {copyright_holder} nor the names of its
+#   contributors may be used to endorse or promote products derived
+#   from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import contextlib
+import functools
+import re
+import unittest
+import yaml
+
+import geometry_msgs.msg
+
+import launch
+import launch.actions
+import launch.substitutions
+import launch_ros.actions
+import launch_ros.substitutions
+import launch_testing.actions
+import launch_testing.tools
+
+from rosidl_runtime_py import message_to_yaml
+
+
+def generate_test_description():
+    path_to_cmd_vel_mux_params = launch.substitutions.PathJoinSubstitution([
+        launch_ros.substitutions.FindPackageShare(package='cmd_vel_mux'),
+        'config',
+        'cmd_vel_mux_params.yaml'
+    ])
+    return launch.LaunchDescription([
+        launch_ros.actions.Node(
+            package='cmd_vel_mux',
+            node_executable='cmd_vel_mux_node',
+            parameters=[path_to_cmd_vel_mux_params],
+            output='screen'
+        ),
+        launch_testing.actions.ReadyToTest()
+    ])
+
+
+class TestCmdVelMux(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(
+        cls,
+        launch_service,
+        proc_info,
+        proc_output
+    ):
+        @contextlib.contextmanager
+        def launch_topic_command(self, arguments):
+            topic_command_action = launch.actions.ExecuteProcess(
+                cmd=['ros2', 'topic', *arguments],
```

Comment:
I'm wondering if using `ros2 topic` like this is brittle.  In particular, if the output format for `ros2 topic` ever changes, then these tests will all start failing.  In some other places (like https://github.com/ros-teleop/teleop_tools/pull/49), I've been constructing the node right in the test and doing the pub/sub by hand there.  Thoughts either way on this?

Reply:
> In particular, if the output format for ros2 topic ever changes, then these tests will all start failing.

That is true. The main reason why I did things this way is to keep tests as simple as possible. I'm wary about tests where support logic is larger / more complex than the test itself. And firing up two publishers at fixed rates while waiting on a specific sequence of messages on subscriptions using `rclpy` will have that effect.

But yeah, nothing is preventing us from rolling out extra testing infrastructure to not rely on `ros2` CLI tools.

Reply:
I tend to prefer having the test code actually be code, rather than fork/exec.  It can give you better error messages on failure, and gives you more fine-grained control.

But if you think that is going to be too much infrastructure/work, I'm fine with leaving this as-is.

Reply:
Fair enough, see https://github.com/clalancette/cmd_vel_mux/pull/8/commits/0bea3e313a5d0377ef6043c1ebc4d4b1ee1dde4b.